File: libs/language/src/lib/merge/__tests__/ListProcessor.test.ts
import { describe, it, expect } from 'vitest';
import { BlueNode } from '../../model';
import { Merger } from '../Merger';
import { MergingProcessor } from '../MergingProcessor';
import { SequentialMergingProcessor } from '../processors/SequentialMergingProcessor';
import { ListProcessor } from '../processors/ListProcessor';
import { TypeAssigner } from '../processors/TypeAssigner';
import { NodeExtender } from '../../utils/NodeExtender';
import { NO_LIMITS } from '../../utils/limits';
import {
  LIST_TYPE_BLUE_ID,
  CORE_TYPE_BLUE_ID_TO_NAME_MAP,
} from '../../utils/Properties';
import { BasicNodeProvider } from '../../provider/BasicNodeProvider';

describe('ListProcessor', () => {
  it('testItemTypeAssignment', () => {
    const nodeProvider = new BasicNodeProvider();

    const listA = new BlueNode()
      .setName('ListA')
      .setType('List')
      .setItemType('Integer');
    nodeProvider.addSingleNodes(listA);

    const listB = new BlueNode()
      .setName('ListB')
      .setType(new BlueNode().setBlueId(nodeProvider.getBlueIdByName('ListA')));
    nodeProvider.addSingleNodes(listB);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new ListProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);

    const listBNode = nodeProvider.findNodeByName('ListB');
    if (!listBNode) {
      throw new Error('ListB not found');
    }

    const result = merger.resolve(listBNode, NO_LIMITS);

    expect(result.getItemType()?.getBlueId()).toBeDefined();
    expect(
      CORE_TYPE_BLUE_ID_TO_NAME_MAP[
        result
          .getItemType()!
          .getBlueId() as keyof typeof CORE_TYPE_BLUE_ID_TO_NAME_MAP
      ],
    ).toBe('Integer');
  });

  it('testListWithValidItemTypes', () => {
    const nodeProvider = new BasicNodeProvider();

    // Create node A
    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    // Create node B
    const b = `name: B
type:
  blueId: ${nodeProvider.getBlueIdByName('A')}`;
    nodeProvider.addSingleDocs(b);

    // Create node C
    const c = `name: C
type:
  blueId: ${nodeProvider.getBlueIdByName('B')}`;
    nodeProvider.addSingleDocs(c);

    // Create ListOfB
    const listOfB = `name: ListOfB
type:
  blueId: ${LIST_TYPE_BLUE_ID}
itemType:
  blueId: ${nodeProvider.getBlueIdByName('B')}
items:
  - type:
      blueId: ${nodeProvider.getBlueIdByName('B')}
  - type:
      blueId: ${nodeProvider.getBlueIdByName('C')}`;
    nodeProvider.addSingleDocs(listOfB);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new ListProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const listOfBNode = nodeProvider.findNodeByName('ListOfB');
    if (!listOfBNode) {
      throw new Error('ListOfB not found');
    }
    new NodeExtender(nodeProvider).extend(listOfBNode, NO_LIMITS);
    const result = merger.resolve(listOfBNode, NO_LIMITS);

    expect(result.getItemType()?.getName()).toBe('B');
    expect(result.getItems()?.length).toBe(2);
    expect(result.getItems()?.[0].getType()?.getName()).toBe('B');
    expect(result.getItems()?.[1].getType()?.getName()).toBe('C');
  });

  it('testListWithInvalidItemType', () => {
    const nodeProvider = new BasicNodeProvider();

    // Create node A
    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    // Create node B
    const b = `name: B
type:
  blueId: ${nodeProvider.getBlueIdByName('A')}`;
    nodeProvider.addSingleDocs(b);

    // Create ListOfB with invalid item type
    const listOfB = `name: ListOfB
type: List
itemType:
  blueId: ${nodeProvider.getBlueIdByName('B')}
items:
  - type:
      blueId: ${nodeProvider.getBlueIdByName('B')}
  - type:
      blueId: ${nodeProvider.getBlueIdByName('A')}`; // This should cause an error
    nodeProvider.addSingleDocs(listOfB);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new ListProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const listOfBNode = nodeProvider.findNodeByName('ListOfB');
    if (!listOfBNode) {
      throw new Error('ListOfB not found');
    }
    new NodeExtender(nodeProvider).extend(listOfBNode, NO_LIMITS);

    expect(() => merger.resolve(listOfBNode, NO_LIMITS)).toThrow();
  });

  it('testInheritedList', () => {
    const nodeProvider = new BasicNodeProvider();

    // Create node A
    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    // Create node B
    const b = `name: B
type:
  blueId: ${nodeProvider.getBlueIdByName('A')}`;
    nodeProvider.addSingleDocs(b);

    // Create node C
    const c = `name: C
type:
  blueId: ${nodeProvider.getBlueIdByName('B')}`;
    nodeProvider.addSingleDocs(c);

    // Create ListOfB
    const listOfB = `name: ListOfB
type:
  blueId: ${LIST_TYPE_BLUE_ID}
itemType:
  blueId: ${nodeProvider.getBlueIdByName('B')}`;
    nodeProvider.addSingleDocs(listOfB);

    // Create InheritedList
    const inheritedList = `name: InheritedList
type:
  blueId: ${nodeProvider.getBlueIdByName('ListOfB')}
items:
  - type:
      blueId: ${nodeProvider.getBlueIdByName('B')}
  - type:
      blueId: ${nodeProvider.getBlueIdByName('C')}`;
    nodeProvider.addSingleDocs(inheritedList);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new ListProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const inheritedListNode = nodeProvider.findNodeByName('InheritedList');
    if (!inheritedListNode) {
      throw new Error('InheritedList not found');
    }
    new NodeExtender(nodeProvider).extend(inheritedListNode, NO_LIMITS);
    const result = merger.resolve(inheritedListNode, NO_LIMITS);

    expect(result.getItemType()?.getName()).toBe('B');
    expect(result.getItems()?.length).toBe(2);
    expect(result.getItems()?.[0].getType()?.getName()).toBe('B');
    expect(result.getItems()?.[1].getType()?.getName()).toBe('C');
  });

  it('testInheritedListWithInvalidItemType', () => {
    const nodeProvider = new BasicNodeProvider();

    // Create node A
    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    // Create node B
    const b = `name: B
type:
  blueId: ${nodeProvider.getBlueIdByName('A')}`;
    nodeProvider.addSingleDocs(b);

    // Create ListOfB
    const listOfB = `name: ListOfB
type:
  blueId: ${LIST_TYPE_BLUE_ID}
itemType:
  blueId: ${nodeProvider.getBlueIdByName('B')}`;
    nodeProvider.addSingleDocs(listOfB);

    // Create InheritedList with invalid item type
    const inheritedList = `name: InheritedList
type:
  blueId: ${nodeProvider.getBlueIdByName('ListOfB')}
items:
  - type:
      blueId: ${nodeProvider.getBlueIdByName('B')}
  - type:
      blueId: ${nodeProvider.getBlueIdByName('A')}`; // This should cause an error
    nodeProvider.addSingleDocs(inheritedList);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new ListProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const inheritedListNode = nodeProvider.findNodeByName('InheritedList');
    if (!inheritedListNode) {
      throw new Error('InheritedList not found');
    }
    new NodeExtender(nodeProvider).extend(inheritedListNode, NO_LIMITS);

    expect(() => merger.resolve(inheritedListNode, NO_LIMITS)).toThrow();
  });

  it('testListWithNoItemType', () => {
    const nodeProvider = new BasicNodeProvider();

    // Create node A
    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    // Create list with no itemType
    const listWithNoItemType = `name: ListWithNoItemType
type:
  blueId: ${LIST_TYPE_BLUE_ID}
items:
  - type:
      blueId: ${nodeProvider.getBlueIdByName('A')}`;
    nodeProvider.addSingleDocs(listWithNoItemType);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new ListProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const listNode = nodeProvider.findNodeByName('ListWithNoItemType');
    if (!listNode) {
      throw new Error('ListWithNoItemType not found');
    }
    new NodeExtender(nodeProvider).extend(listNode, NO_LIMITS);
    const result = merger.resolve(listNode, NO_LIMITS);

    expect(result.getItemType()).toBeUndefined();
    expect(result.getItems()?.length).toBe(1);
    expect(result.getItems()?.[0].getType()?.getName()).toBe('A');
  });

  it('testNonListTypeWithItemType', () => {
    const nodeProvider = new BasicNodeProvider();

    // Create node A
    const a = 'name: A';
    nodeProvider.addSingleDocs(a);

    // Create non-list with itemType
    const nonListWithItemType = `name: NonListWithItemType
type:
  blueId: ${nodeProvider.getBlueIdByName('A')}
itemType:
  blueId: ${nodeProvider.getBlueIdByName('A')}`;

    nodeProvider.addSingleDocs(nonListWithItemType);

    const mergingProcessor: MergingProcessor = new SequentialMergingProcessor([
      new TypeAssigner(),
      new ListProcessor(),
    ]);

    const merger = new Merger(mergingProcessor, nodeProvider);
    const nonListNode = nodeProvider.findNodeByName('NonListWithItemType');
    if (!nonListNode) {
      throw new Error('NonListWithItemType not found');
    }
    new NodeExtender(nodeProvider).extend(nonListNode, NO_LIMITS);

    expect(() => merger.resolve(nonListNode, NO_LIMITS)).toThrow();
  });
});

File: references/java-sdk/src/test/java/blue/language/ListProcessorTest.java
package blue.language;

import blue.language.merge.Merger;
import blue.language.merge.MergingProcessor;
import blue.language.model.Node;
import blue.language.merge.processor.ListProcessor;
import blue.language.merge.processor.SequentialMergingProcessor;
import blue.language.merge.processor.TypeAssigner;
import blue.language.provider.BasicNodeProvider;
import blue.language.utils.NodeExtender;
import blue.language.utils.Properties;
import blue.language.utils.limits.Limits;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static blue.language.blueid.legacy.LegacyBlueIdCalculator.calculateBlueId;
import static blue.language.utils.Properties.CORE_TYPE_BLUE_ID_TO_NAME_MAP;
import static org.junit.jupiter.api.Assertions.*;

public class ListProcessorTest {

    @Test
    public void testItemTypeAssignment() {
        Node listA = new Node().name("ListA")
                .type("List")
                .itemType("Integer");
        Node listB = new Node().name("ListB")
                .type(new Node().blueId(calculateBlueId(listA)));

        List<Node> nodes = Arrays.asList(listA, listB);
        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new ListProcessor()
                )
        );

        BasicNodeProvider nodeProvider = new BasicNodeProvider(nodes);
        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node listANode = nodeProvider.findNodeByName("ListA").orElseThrow(() -> new IllegalStateException("No \"ListA\" available for NodeProvider."));
        Node result = merger.resolve(listANode, Limits.NO_LIMITS);

        assertEquals("Integer", CORE_TYPE_BLUE_ID_TO_NAME_MAP.get(result.getItemType().getBlueId()));
    }

    @Test
    public void testListWithValidItemTypes() throws Exception {

        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String b = "name: B\n" +
                   "type:\n" +
                   "  blueId: " + nodeProvider.getBlueIdByName("A");
        nodeProvider.addSingleDocs(b);

        String c = "name: C\n" +
                   "type:\n" +
                   "  blueId: " + nodeProvider.getBlueIdByName("B");
        nodeProvider.addSingleDocs(c);

        String listOfB = "name: ListOfB\n" +
                         "type:\n" +
                         "  blueId: " + Properties.LIST_TYPE_BLUE_ID + "\n" +
                         "itemType:\n" +
                         "  blueId: " + nodeProvider.getBlueIdByName("B") + "\n" +
                         "items:\n" +
                         "  - type:\n" +
                         "      blueId: " + nodeProvider.getBlueIdByName("B") + "\n" +
                         "  - type:\n" +
                         "      blueId: " + nodeProvider.getBlueIdByName("C");
        nodeProvider.addSingleDocs(listOfB);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new ListProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node listOfBNode = nodeProvider.getNodeByName("ListOfB");
        new NodeExtender(nodeProvider).extend(listOfBNode, Limits.NO_LIMITS);
        Node result = merger.resolve(listOfBNode);

        assertEquals("B", result.getItemType().getName());
        assertEquals(2, result.getItems().size());
        assertEquals("B", result.getItems().get(0).getType().getName());
        assertEquals("C", result.getItems().get(1).getType().getName());
    }

    @Test
    public void testListWithInvalidItemType() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String b = "name: B\n" +
                   "type: " + nodeProvider.getBlueIdByName("A");
        nodeProvider.addSingleDocs(b);

        String listOfB = "name: ListOfB\n" +
                         "type: List\n" +
                         "itemType: " + nodeProvider.getBlueIdByName("B") + "\n" +
                         "items:\n" +
                         "  - type: " + nodeProvider.getBlueIdByName("B") + "\n" +
                         "  - type: " + nodeProvider.getBlueIdByName("A");  // This should cause an error
        nodeProvider.addSingleDocs(listOfB);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new ListProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node listOfBNode = nodeProvider.findNodeByName("ListOfB").orElseThrow(() -> new IllegalStateException("No \"ListOfB\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(listOfBNode, Limits.NO_LIMITS);

        assertThrows(IllegalArgumentException.class, () -> merger.resolve(listOfBNode));
    }

    @Test
    public void testInheritedList() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String b = "name: B\n" +
                   "type:\n" +
                   "  blueId: " + nodeProvider.getBlueIdByName("A");
        nodeProvider.addSingleDocs(b);

        String c = "name: C\n" +
                   "type:\n" +
                   "  blueId: " + nodeProvider.getBlueIdByName("B");
        nodeProvider.addSingleDocs(c);

        String listOfB = "name: ListOfB\n" +
                         "type:\n" +
                         "  blueId: " + Properties.LIST_TYPE_BLUE_ID + "\n" +
                         "itemType:\n" +
                         "  blueId: " + nodeProvider.getBlueIdByName("B");
        nodeProvider.addSingleDocs(listOfB);

        String inheritedList = "name: InheritedList\n" +
                               "type:\n" +
                               "  blueId: " + nodeProvider.getBlueIdByName("ListOfB") + "\n" +
                               "items:\n" +
                               "  - type:\n" +
                               "      blueId: " + nodeProvider.getBlueIdByName("B") + "\n" +
                               "  - type:\n" +
                               "      blueId: " + nodeProvider.getBlueIdByName("C");
        nodeProvider.addSingleDocs(inheritedList);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new ListProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node inheritedListNode = nodeProvider.findNodeByName("InheritedList").orElseThrow(() -> new IllegalStateException("No \"InheritedList\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(inheritedListNode, Limits.NO_LIMITS);
        Node result = merger.resolve(inheritedListNode);

        assertEquals("B", result.getItemType().getName());
        assertEquals(2, result.getItems().size());
        assertEquals("B", result.getItems().get(0).getType().getName());
        assertEquals("C", result.getItems().get(1).getType().getName());
    }

    @Test
    public void testInheritedListWithInvalidItemType() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String b = "name: B\n" +
                   "type:\n" +
                   "  blueId: " + nodeProvider.getBlueIdByName("A");
        nodeProvider.addSingleDocs(b);

        String listOfB = "name: ListOfB\n" +
                         "type:\n" +
                         "  blueId: " + Properties.LIST_TYPE_BLUE_ID + "\n" +
                         "itemType:\n" +
                         "  blueId: " + nodeProvider.getBlueIdByName("B");
        nodeProvider.addSingleDocs(listOfB);

        String inheritedList = "name: InheritedList\n" +
                               "type:\n" +
                               "  blueId: " + nodeProvider.getBlueIdByName("ListOfB") + "\n" +
                               "items:\n" +
                               "  - type:\n" +
                               "      blueId: " + nodeProvider.getBlueIdByName("B") + "\n" +
                               "  - type:\n" +
                               "      blueId: " + nodeProvider.getBlueIdByName("A");  // This should cause an error
        nodeProvider.addSingleDocs(inheritedList);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new ListProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node inheritedListNode = nodeProvider.findNodeByName("InheritedList").orElseThrow(() -> new IllegalStateException("No \"InheritedList\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(inheritedListNode, Limits.NO_LIMITS);

        assertThrows(IllegalArgumentException.class, () -> merger.resolve(inheritedListNode));
    }

    @Test
    public void testListWithNoItemType() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String listWithNoItemType = "name: ListWithNoItemType\n" +
                                    "type:\n" +
                                    "  blueId: " + Properties.LIST_TYPE_BLUE_ID + "\n" +
                                    "items:\n" +
                                    "  - type:\n" +
                                    "      blueId: " + nodeProvider.getBlueIdByName("A");
        nodeProvider.addSingleDocs(listWithNoItemType);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new ListProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node listNode = nodeProvider.findNodeByName("ListWithNoItemType").orElseThrow(() -> new IllegalStateException("No \"ListWithNoItemType\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(listNode, Limits.NO_LIMITS);
        Node result = merger.resolve(listNode);

        assertNull(result.getItemType());
        assertEquals(1, result.getItems().size());
        assertEquals("A", result.getItems().get(0).getType().getName());
    }

    @Test
    public void testNonListTypeWithItemType() throws Exception {
        BasicNodeProvider nodeProvider = new BasicNodeProvider();

        String a = "name: A";
        nodeProvider.addSingleDocs(a);

        String nonListWithItemType = "name: NonListWithItemType\n" +
                                     "type: " + nodeProvider.getBlueIdByName("A") + "\n" +
                                     "itemType: " + nodeProvider.getBlueIdByName("A");
        nodeProvider.addSingleDocs(nonListWithItemType);

        MergingProcessor mergingProcessor = new SequentialMergingProcessor(
                Arrays.asList(
                        new TypeAssigner(),
                        new ListProcessor()
                )
        );

        Merger merger = new Merger(mergingProcessor, nodeProvider);
        Node nonListNode = nodeProvider.findNodeByName("NonListWithItemType").orElseThrow(() -> new IllegalStateException("No \"NonListWithItemType\" available for NodeProvider."));
        new NodeExtender(nodeProvider).extend(nonListNode, Limits.NO_LIMITS);

        assertThrows(IllegalArgumentException.class, () -> merger.resolve(nonListNode));
    }
}